RecoEgamma/EgammaIsolationAlgos/plugins/isolation_cones/PhotonPFIsolationWithMapBasedVeto.cc

```diff
@@ -0,0 +1,170 @@
+{
+    for ( auto itr = thefootprint.begin(); itr != thefootprint.end(); ++itr ) 
+    {
+      if( itr->key() == theCandidate->sourceCandidatePtr(0).key() ) return true;
+    }
+    return false;
+}
+
+class PhotonPFIsolationWithMapBasedVeto : public citk::IsolationConeDefinitionBase {
+public:
+  PhotonPFIsolationWithMapBasedVeto(const edm::ParameterSet& c) :
+    citk::IsolationConeDefinitionBase(c),
+    _isolateAgainst(c.getParameter<std::string>("isolateAgainst")), //isolate against either h+, h0 or gamma
+    _miniAODVertexCodes(c.getParameter<std::vector<unsigned> >("miniAODVertexCodes")), //quality flags to be used for association with the vertex configurable, the vertex can be chosen
+    _vertexIndex(c.getParameter<int> ("vertexIndex")){} //vertex of interest
+        
+  PhotonPFIsolationWithMapBasedVeto(const PhotonPFIsolationWithMapBasedVeto&) = delete;
+  PhotonPFIsolationWithMapBasedVeto& operator=(const PhotonPFIsolationWithMapBasedVeto&) =delete;
+
+  bool isInIsolationCone(const reco::CandidatePtr& photon,
+			 const reco::CandidatePtr& PFCandidate) const override final;
+  
+   
+  // this object is needed for reco case
+ edm::Handle< edm::ValueMap<std::vector<reco::PFCandidateRef > > > particleBasedIsolationMap;			 
+ edm::EDGetTokenT<edm::ValueMap<std::vector<reco::PFCandidateRef > > > particleBasedIsolationToken_;
+ 
+ virtual void getEventInfo(const edm::Event& iEvent)
+  {
+      iEvent.getByToken(particleBasedIsolationToken_, particleBasedIsolationMap); 
+  };			 
+  
+  
+  //As far as I understand now, the object particleBasedIsolationMap should be fixed, so we don't configure the name
+  void setConsumes(edm::ConsumesCollector iC)
+  {
+      particleBasedIsolationToken_ = iC.mayConsume<edm::ValueMap<std::vector<reco::PFCandidateRef > > >(edm::InputTag("particleBasedIsolation", "gedPhotons"));
+  }
+
+  //! Destructor
+  virtual ~PhotonPFIsolationWithMapBasedVeto(){};
+  
+  
+private:    
+  const std::string _isolateAgainst, _vertexCollection;
+  const std::vector<unsigned> _miniAODVertexCodes;
+  const unsigned _vertexIndex;
+
+  
+};
+
+DEFINE_EDM_PLUGIN(CITKIsolationConeDefinitionFactory,
+		  PhotonPFIsolationWithMapBasedVeto,
+		  "PhotonPFIsolationWithMapBasedVeto");
+
+
+//This function defines whether particular PFCandidate is inside of isolation cone of photon or not by checking deltaR and whether footprint removal for this candidate should be done. Additionally, for miniAOD charged hadrons from the PV are considered. *** For AOD this should be done by the corresponding sequence beforehand!!! ***
+ 
+bool PhotonPFIsolationWithMapBasedVeto::isInIsolationCone(const reco::CandidatePtr& photon,  const reco::CandidatePtr& PFCandidate  ) const {
+ 
+  //convert the photon and candidate objects to the corresponding pat or reco objects. What is used depends on what is user running on: miniAOD or AOD
+  pat::patPhotonPtr aspat_photonptr(photon);
+  
+  pat::PackedCandidatePtr aspacked(PFCandidate);
```

Comment:
to clarify a bit: "Instead of using persisted Ptr classes, use plain C++ " means  instead of pat::PackedCandidatePtr to use "pat::PackedCandidate const*"